x-pack/plugins/apm/public/components/alerting/ui_components/apm_rule_kql_filter/autocomplete_field.tsx

```diff
@@ -0,0 +1,344 @@
+      selectedSuggestion.text +
+      value.substr(selectedSuggestion.end);
+
+    this.setState(withSuggestionsHidden);
+    this.changeValue(newValue);
+    this.focusInputElement();
+  };
+
+  private changeValue = (value: string) => {
+    const { onChange } = this.props;
+
+    if (onChange) {
+      onChange(value);
+    }
+  };
+
+  private focusInputElement = () => {
+    if (this.inputElement) {
+      this.inputElement.focus();
+    }
+  };
+
+  private showSuggestions = () => {
+    this.setState(withSuggestionsVisible);
+  };
+
+  private submit = () => {
+    const { isValid, onSubmit, value } = this.props;
+
+    if (isValid && onSubmit) {
+      onSubmit(value);
+    }
+
+    this.setState(withSuggestionsHidden);
+  };
+
+  private updateSuggestions = () => {
+    const inputCursorPosition = this.inputElement
+      ? this.inputElement.selectionStart || 0
+      : 0;
+    this.props.loadSuggestions(this.props.value, inputCursorPosition, 200);
+  };
+}
+
+const withPreviousSuggestionSelected = (
+  state: AutocompleteFieldState,
+  props: AutocompleteFieldProps
+): AutocompleteFieldState => ({
+  ...state,
+  selectedIndex:
+    props.suggestions.length === 0
+      ? null
+      : state.selectedIndex !== null
+      ? (state.selectedIndex + props.suggestions.length - 1) %
+        props.suggestions.length
+      : Math.max(props.suggestions.length - 1, 0),
+});
+
+const withNextSuggestionSelected = (
+  state: AutocompleteFieldState,
+  props: AutocompleteFieldProps
+): AutocompleteFieldState => ({
+  ...state,
+  selectedIndex:
+    props.suggestions.length === 0
+      ? null
+      : state.selectedIndex !== null
+      ? (state.selectedIndex + 1) % props.suggestions.length
+      : 0,
+});
+
+const withSuggestionAtIndexSelected =
+  (suggestionIndex: number) =>
+  (
+    state: AutocompleteFieldState,
+    props: AutocompleteFieldProps
+  ): AutocompleteFieldState => ({
+    ...state,
+    selectedIndex:
+      props.suggestions.length === 0
+        ? null
+        : suggestionIndex >= 0 && suggestionIndex < props.suggestions.length
+        ? suggestionIndex
+        : 0,
+  });
+
+const withSuggestionsVisible = (state: AutocompleteFieldState) => ({
+  ...state,
+  areSuggestionsVisible: true,
+});
+
+const withSuggestionsHidden = (state: AutocompleteFieldState) => ({
+  ...state,
+  areSuggestionsVisible: false,
+  selectedIndex: null,
+});
+
+const withFocused = (state: AutocompleteFieldState) => ({
+  ...state,
+  isFocused: true,
+});
+
+const withUnfocused = (state: AutocompleteFieldState) => ({
+  ...state,
+  isFocused: false,
+});
+
+const AutocompleteContainer = euiStyled.div`
+  position: relative;
+`;
+
+const SuggestionsPanel = euiStyled(EuiPanel).attrs(() => ({
+  paddingSize: 'none',
+  hasShadow: true,
+}))`
+  position: absolute;
+  width: 100%;
+  margin-top: 2px;
+  overflow-x: hidden;
+  overflow-y: scroll;
+  z-index: ${(props) => props.theme.eui.euiZLevel1};
+  max-height: 322px;
+`;
```

Comment:
Hmm... Does this component not already exist somewhere? Is this not something Unified search can help us with? (@achyutjhunjhunwala worked on unified search and might have an idea)

Reply:
Refactored the code to reuse KQL filter related components from `observability` plugin. I looked into Unified Search component, but it has some additional functionalities like adding params to URL, calendar, refresh, etc. that we don't need for adding filter query in rule.

Reply:
@benakansara The UnifiedSearch component provided by the Kibana team has all the these functions exposed as props. Which means you can use the search bar and pass your own functions to `handleSubmit` function. It also accept props to show/hide, calendar, submit button etc. The APM version uses the Unified Search and passes its own handlers. 

https://github.com/elastic/kibana/blob/75140a89775a4341cdbd8d0788971b3fde1fb010/x-pack/plugins/apm/public/components/shared/unified_search_bar/index.tsx#L285-L305

This way you will save effort in creating the boiler plate of typeahead component and styling and feature wise it will close to a searchbar which is used everywhere else in Kibana

Reply:
@achyutjhunjhunwala Thank you for the info. Indeed it would be good to use UnifiedSearch component. I will do this in a follow up PR. Here is the issue for it: https://github.com/elastic/kibana/issues/163781